### PR TITLE
Add DataflowBlockOptions.ForceOrdered to Dataflow library

### LINKED ISSUE #536: Add option to allow for unordered results from TransformBlock
I understand the design intention for having TransformBlock return items in the order they were added where parallel execution is involved. This is great when ordering is important but is a hindrance when ordering is not important.

For example, I want to push new items into TransformBlock and have them start processing as soon as others are completed. In my scenario there is no relationship between each item, but they do require a transform. 

Consider 10 items being added (MaxDegreeOfParallelism = 10) where the 5th takes much longer to process than the others. A further 10 items can then be added and buffered. No items after the 10th will be processed until the 5th has completed therefore allowing 6-10 to be completed.

If output ordering could be disabled via options, then TransformBlock could start processing 11-20 because 6-10 have completed even though the 5th has not yet completed.

### PR DESCRIPTION
**Pull request by stephentoub**
Adds the DataflowBlockOptions.ForceOrdered option, and utilizes it from TransformBlock and TransformManyBlock.

Fixes #536 
cc: @AlfredoMS, @mellinoe 

@ericstj, is there anything I need to do around assembly versionining?

### CODE REVIEW

**Comment by ericstj** [comment]
Since you are adding a new API since the previous stable release you should update the minor version of the assembly.

File: src/System.Threading.Tasks.Dataflow/src/Base/DataflowBlockOptions.cs [hunk 1]
```diff
@@ -170,6 +177,13 @@ public string NameFormat
                 _nameFormat = value;
             }
         }
+
+        /// <summary>Gets whether ordered processing should be enforced on a block's handling of messages.</summary>
```
Comment (mellinoe) [inline src/System.Threading.Tasks.Dataflow/src/Base/DataflowBlockOptions.cs]: "Gets or sets"

Comment (ericstj) [inline src/System.Threading.Tasks.Dataflow/src/Base/DataflowBlockOptions.cs]: Can you be a bit clearer in the docs:
What order?  When this is false what is the expected behavior?  What's the tradeoff for setting this one way or the other.
"Force" seems to indicate that the default is false when in fact it is true.  Perhaps "ProcessInOrder" might be a better name?

Comment (mellinoe) [inline src/System.Threading.Tasks.Dataflow/src/Base/DataflowBlockOptions.cs]: I wonder if we should "invert" the option, i.e. `AllowUnorderedProcessing` or something like that? I usually prefer configurable options to default to false, but that might just be a personal preference thing.

**Comment by stephentoub** [comment]
> Since you are adding a new API since the previous stable release you should update the minor version of the assembly.

Thanks.  Do I leave the build and revision as-is, or do I set them to 0?

File: src/System.Threading.Tasks.Dataflow/src/Base/DataflowBlockOptions.cs (hunk 1 above)
Comment (stephentoub) [inline src/System.Threading.Tasks.Dataflow/src/Base/DataflowBlockOptions.cs]: > "Gets or sets"

Done.

> Can you be a bit clearer in the docs

Done.

> I wonder if we should "invert" the option, i.e. AllowUnorderedProcessing or something like that?

I understand that preference.  And in general I agree with you.  However, in this case it felt like a double-negative.  Additionally, the library has been around for a while and ordering is effectively always on; this new option provides a way to effectively turn it off, hence defaulting to true and allowing it to be set to false.

That said, while I prefer the default of true in this case, I don't feel terribly strong about it and would be ok inverting it if we can come up with a good name for it and if others prefer the inversion.

File: src/System.Threading.Tasks.Dataflow/src/System.Threading.Tasks.Dataflow.csproj [hunk 2]
```diff
@@ -6,7 +6,7 @@
     <AppDesignerFolder>Properties</AppDesignerFolder>
     <RootNamespace>System.Threading.Tasks.Dataflow</RootNamespace>
     <AssemblyName>System.Threading.Tasks.Dataflow</AssemblyName>
-    <AssemblyVersion>4.5.26.0</AssemblyVersion>
+    <AssemblyVersion>4.6.26.0</AssemblyVersion>
```
Comment (AlfredoMS) [inline src/System.Threading.Tasks.Dataflow/src/System.Threading.Tasks.Dataflow.csproj]: when changing minor version, shouldn't that reset the next number too.
4.5.26.0 -> 4.6.0.0 ?
@ericstj

Comment (ericstj) [inline src/System.Threading.Tasks.Dataflow/src/System.Threading.Tasks.Dataflow.csproj]: Yes

**Comment by stephentoub** [comment]
Any other feedback?  @terrajobst, we're good-to-go with the API name, right?

**Comment by mellinoe** [comment]
The code changes LGTM. I'm fine with the current property name, just wanted to throw out an alternative option. Your reasoning makes sense to me.

**Comment by ericstj** [comment]
I was OK with a `true` default but it seemed to me that `Force` implied that the default was false and you'd set it to true to force something.  That's why I suggested `ProcessInOrder` since it doesn't imply a default of true or false.  Regardless I don't feel strongly about this.  Just sharing one devs opinion before I read the docs.

**Comment by stephentoub** [comment]
Other name options:
- GuaranteeOrdering
- RequireOrdering
etc.  I'm open to other names, too, if folks don't like ForceOrdered.

**Comment by i3arnon** [comment]
I was initially surprised that the default was `true` (though it's clear why that is).
Maybe `EnableUnordered` with `false` as a default.
This implies a `false` default and setting `true` makes it clear that order may not be kept.

**Comment by stephentoub** [comment]
This might just be my own mental hang up, but I think of "un"ordered as a negative, and so EnableUnordered==false as a double-negative that means ordered, and that takes me longer to reason about.  Maybe it's just me.

**Comment by i3arnon** [comment]
>  ... saying "enable unordered" suggests more of a guarantee that things will be unordered

I agree. How about `AllowUnordered`? (`ForceOrdered` seems to suggest that this enables something in a higher magnitude than simply `Ordered` e.g. "force quit", "push --force" which isn't the case here)

>  I think of "un"ordered as a negative

I get that in the language sense, but in a technical (and maybe physical) sense I see "unordered" not as the opposite of "ordered" but as the baseline natural state where ordering requires energy to achieve and preserve.

**Comment by stephentoub** [comment]
> in a technical (and maybe physical) sense I see "unordered" not as the opposite of "ordered" but as the baseline natural state where ordering requires energy to achieve and preserve

In which case it makes sense that the "unordered" state would be false and the "ordered" state true, as according to that definition it's something extra that essentially needs to be enabled.  It just so happens that the initial design/implementation of the library enables it, which is why this property would then default to true, so as not to be a breaking change.

> ForceOrdered seems to suggest that this enables something in a higher magnitude

That's why I suggested alternate prefixes, e.g. EnsureOrdered, GuaranteeOrdered, etc.  EnsureOrdered?

**Comment by stephentoub** [comment]
@terrajobst, opinion on the name?

**Comment by stephentoub** [comment]
I'm going to change it to EnsureOrdered.  Please let me know asap if there are any objections.

**Comment by stephentoub** [comment]
@dotnet-bot test this please

**Comment by mellinoe** [comment]
EnsureOrdered sounds fine to me.

**Pull request merged**